server/aesdsocket.c

```diff
@@ -177,150 +362,242 @@ int main(int argc, char *argv[])
-	
 
-		//Signals from set are added to invoking process's signal mask
-        if (sigprocmask(SIG_BLOCK, &mask, NULL)) 
+		datap = malloc(sizeof(slist_data_t));
+		if (datap == NULL)
 		{
-            syslog(LOG_ERR, "sigprocmask");
-			closeall();
-            exit(EXIT_FAILURE);
-        }
-
-		do
-		{
-			memset(buffer, 0, sizeof(buffer));
-			
-			nbytes = recv(acceptedfd, buffer, sizeof(buffer), 0);
-
-			if (nbytes)
-			{
-				// syslog(LOG_DEBUG, "nbytes - %d\n", nbytes);
-				// syslog(LOG_DEBUG, "'%s'\n", buffer);
-				if(nbytes > strlen(buffer))
-					nr = write(fd, buffer, strlen(buffer));
-				else
-					nr = write(fd, buffer, nbytes);
-					
-				if (nr == -1)	
-				{//if error
-					syslog(LOG_ERR, "can't write received string in file '%s'\n", filepath);
-					break;	
-				}
-				total_bytes += nr;
-			}
-			else
-				break;
-
-		}while(strchr(buffer, '\n') == NULL);
-		
-				
-		nbytes=0;
-		lseek(fd, 0, SEEK_SET);
-		
-		sendbuffer = (char *)malloc(sizeof(char) * (total_bytes+location));
-		if (sendbuffer == NULL)
 			syslog(LOG_ERR, "malloc failed\n");
-		
-		memset(sendbuffer, 0, total_bytes+location);
-		nr = read(fd, sendbuffer, total_bytes+location);
-		if (nr != total_bytes+location)	
-		{//if error
-			syslog(LOG_ERR, "can't read proper bytes from file '%s'\n", filepath);
-			break;	
+			exit_on_error = true;
+			goto EXITING;
 		}
-		
-		//syslog(LOG_DEBUG, "'%s'\n", sendbuffer);
-		nbytes = send(acceptedfd, sendbuffer, nr, 0);
-		if(nbytes != nr)
+			
+		SLIST_INSERT_HEAD(&head,datap,entries);
+		datap->params.acceptedfd = acceptedfd;
+		datap->params.complete_status_flag = false;
+		datap->params.thread_complete_success = false;
+		datap->params.sin_addr = saddr.sin_addr;
+		datap->params.fd=fd;
+
+		pthread_create(&(datap->params.thread), NULL, (void*)&packetRWthread, (void*)&(datap->params));
+
+		SLIST_FOREACH(datap,&head,entries)
 		{
-			syslog(LOG_ERR, "not all bytes sent\n");
-			break;
-		}
-		//syslog(LOG_DEBUG, "nbytes - %d\n", nbytes);
-		//total_bytes -= nbytes;
-		total_bytes = 0;
-		
-		location=lseek(fd, 0, SEEK_END);
+		    if (datap->params.complete_status_flag == true)
+		        pthread_join(datap->params.thread,NULL);
+		    else
+		    	continue;
+    	}
+	}
 
-		syslog(LOG_DEBUG, "Closed connection from '%s'\n", inet_ntoa((struct in_addr)saddr.sin_addr));
+EXITING:
 
+	if (exit_on_error && exit_on_signal==false)
+		syslog(LOG_DEBUG, "exiting on failure\n");
+	else
+		syslog(LOG_DEBUG, "exiting on signal\n");
 
-		//Signals from set are added to invoking process's signal mask
-        if (sigprocmask(SIG_UNBLOCK, &mask, NULL)) 
-		{
-            syslog(LOG_ERR, "sigprocmask");
-			closeall();
-            exit(EXIT_FAILURE);
-        }
+	//if(acceptedfd)
+	//	close(acceptedfd);
+
+	if(sockfd)
+		close(sockfd);
+
+	if(fd)	
+	{
+		close(fd);
+		remove(DEF_FILEPATH);
+	}
+	
+	pthread_mutex_destroy(&mutex);
```

Comment:
@MuktaDarekar what could happen if you `mutex_destroy` before `pthread_join` and one of your threads attempts to access `mutex`?  This might be wy you needed pthread_cancel below.  This shouldn't be necessary.